[FLINK-2457] Integrate Tuple0

mjsax opened the pull request:
- extended TypeExtractor to handle Tuple0
 - extended .getTupleClass(int) to hanlde arity zero
 - simplified Receiver.createTuple(int)

StephanEwen commented:
I think this patch needs to include the proper serializers and comparators for the tuple0. Otherwise it now allows users to use the type, but the runtime may fail on it, which is a dangerous inconsistent state.

mjsax commented:
I double checked with two different programs and it worked... Not sure if I am missing something. Please see dev mailing list. Otherwise, please help me to come up with an example that breaks serialization such that I can use this example to fix the code. I did not try comparators though: not sure for what this would be helpful? There is no data in Tuple0 so there is nothing to compare with. IMHO, all Tuple0 object are equal to each other. I don't see any use case to group on Tuple0, thus, comparators are not required from my point of view and should not be supported.

StephanEwen commented:
Are you sending the Tuple0 as part of another job, or alone?

Try the following: Job that generates n Tuple0, write it with a TypeSerializerOutputFormat. Then read it back in with a TypeSerializerInputFormat.

mjsax commented:
I used different dop for source/consumer and double check that multiple TMs are involved to make sure data is really going over the network...

Batch example:
```
public class TestEmptyTupleBatch {

	public static void main(String[] args) throws Exception {
		final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

		DataSource<Tuple0> input = env.fromElements(
				new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(),
				new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(),
				new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(),
				new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0(), new Tuple0());

		input.map(new Counter()).setParallelism(4).print();
	}

	public static class Counter implements MapFunction<Tuple0, Integer> {
		private static final long serialVersionUID = 5518823010274195186L;

		int counter = 0;
		@Override
		public Integer map(Tuple0 value) throws Exception {
			System.out.println("mjsax: " + (++counter) + " " + value.getField(0));
			return new Integer(value.getArity());
		}
	}
}
```

Streaming Example:
```
public class TestEmptyTuple {

	public static void main(String[] args) throws Exception {
		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

		DataStream<Tuple0> input = env.addSource(new Empty()).setParallelism(4);
		input.shuffle().transform("verify", null, new Counter()).setParallelism(1);

		env.execute();
	}

	public static class Empty implements ParallelSourceFunction<Tuple0> {
		private static final long serialVersionUID = 1350902748274781033L;
		private volatile boolean isRunning = true;

		@Override
		public void run(org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext<Tuple0> ctx)
						throws Exception {
			while (this.isRunning) {
				Thread.sleep(1000);
				ctx.collect(new Tuple0());
			}
		}

		@Override
		public void cancel() {
			this.isRunning = false;
		}
	}

	public static class Counter implements OneInputStreamOperator<Tuple0,Object> {
		private static final long serialVersionUID = 5518823010274195186L;

		private int counter = 0;
		@Override
		public void processElement(StreamRecord<Tuple0> element) throws Exception {
			System.out.println("Tuple0: " + (++counter));
		}

		// 
	}

}
```

StephanEwen commented:
Try the example I wrote. I am pretty sure that one will not work.

StephanEwen commented:
In many cases, the problem that Tuple0 does not consume anything, will be caught by the fact that other data that is stored or shipped with the tuple will advance the stream.

However, relying on that is super dangerous. As soon as you hit a case where that is not given, it will not work any more.

twalthr commented:
I think this PR will introduce multiple bugs.
@mjsax please also have a look into the `TypeInfoParser`.

twalthr commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:
```diff
@@ -391,6 +392,10 @@ protected TypeExtractor() {
 				curT = typeToClass(curT).getGenericSuperclass();
 			}
 			
+			if(typeToClass(curT).equals(Tuple0.class)) {
```
`curT` could be `TypeVariable` or someting else at this point. I think it's better to use `if (curT instanceof Class<?> && curT == Tuple0.class)`

mjsax replied:
Yes it can. But in this case we are not done and do not return here. The test for `if (curT instanceof Class<?>)` is done below... This `if` is an early return that is only valid if the type is `Tuple0`. Or do I miss something?

twalthr replied:
No, calling `typeToClass()` with a TypeVar will throw an `IllegalArgumentException`. `typeToClass()` should only be used if  `isClassType()` is true.

mjsax replied:
I see. Makes sense. I will update it.

mjsax commented:
@twalthr thanks for the hint to `TypeInfoParser`. I will have a lock. But your statement about "introduce multiple bugs", seems to be a little bit to general... I did not break any tests... If there is untested code that I might have broken, please let me know. For this case, we should extend test coverage for those classes.

twalthr commented:
Sorry, I was just thinking loadly. You are right ;)

mjsax commented:
I just tried using `TypeSerializerOutputFormat` for a DataSet sink. There is no error. The system only creates an empty file. How can I force the usage of a file-channel (instead of network or in-memory chaining) between two mappers?

fhueske commented:
The empty file is exactly the problem. 
What happens if you read that file with a TypeSerializerInpitFormat typed for Tuple0? It will not return the same amount of Tuple0 objects as were "written" to it. That's the problem with de/serializers that do not read/write anything. This problem can occur anywhere in Flink's engine where data is serialized.

fhueske commented:
I tried to write a DataSet program that processes Tuple0 records and fails during execution. However, this is surprisingly hard. In fact, I didn't manage to break the system.

The reason why we argued that Tuple0 is unsafe is because de/serialization does not read/write anything and the byte stream is not forwarded. Hence you could "read" a million Tuple0 objects without advancing the stream. This is only relevant for DataSet that consist completely of types that have this behavior, because as soon there is another type with proper de/serialization the stream is advanced. 

I was surprised when @mjsax said, that a job that shuffled Tuple0 records (which means de/serialization) worked. I verified that and my guess would be that the network stack writes the number of records into its network buffer and stops deserializing after all records are deserialized (not when the byte stream is EOF). Hence, network serialization works even if no bytes are shipped. @uce or @StephanEwen might confirm this.

Another reason to de/serialize data is for sorting, grouping, joining, or crossing. Most transformation that require a key (there are some inconsistency wrt to key handling in the API) do not allow Tuple0 keys and fail with a good error message when the program is constructed (not at runtime). The API does not allow to sort, join, or group a data set on a Tuple0. I thought that using a cross transformation would let a program fail while executing because it does not require a key and should serialize data, however it worked. 

The system seems to be more robust than I expected. Nonetheless, I am still a bit skeptical about this change and would like to learn why network transfer and crossing worked so well.

mjsax commented:
Thanks for conforming my observations and even dive deeper into it. What should the next step be?

StephanEwen commented:
Quoting from above:

In many cases, the problem that Tuple0 does not consume anything, will be caught by the fact that other data that is stored or shipped with the tuple will advance the stream. However, relying on that is super dangerous. As soon as you hit a case where that is not given, it will not work any more.

Not being able to find a counter example does not mean it is correct. 

The network stack works, because it ships metadata bytes per element. With the TypeSerializerInput/Output format example, the output writes for 10 elements an empty file, the input reads either none or infinitely many. There is at least one counterexample.

What is so bad about creating a `Tuple0Serializer` that writes a dummy byte per tuple?

mjsax commented:
There is nothing bad about it. I just would like to have a breaking example...

mjsax commented:
Added serialization and extended tests. I tested the serialization part by writing Tuple0 as result of one batch program to disc and consumed those Tuple0 recored in a second program successfully. Please give feedback.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/TupleSerializer.java:
```diff
@@ -97,48 +97,63 @@ public T copy(T from) {
-				fieldSerializers[i].serialize(o, target);
-			} catch (NullPointerException npex) {
-				throw new NullFieldException(i, npex);
+		if (arity > 0) {
```
IMO, it would be better if the TupleTypeInfo returns a special Tuple0Serializer instead of doing this check for each record during execution.

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/type/extractor/TypeExtractorTest.java:
```diff
@@ -200,12 +201,12 @@ public void testTupleWithBasicTypes() throws Exception {
 	@Test
 	public void testTupleWithTuples() {
 		// use getFlatMapReturnTypes()
-		RichFlatMapFunction<?, ?> function = new RichFlatMapFunction<Tuple3<Tuple1<String>, Tuple1<Integer>, Tuple2<Long, Long>>, Tuple3<Tuple1<String>, Tuple1<Integer>, Tuple2<Long, Long>>>() {
```
Can you add new tests instead of changing existing ones?

fhueske commented:
Thanks for updating the PR!
Apart from two comments, it looks good to me.

mjsax commented:
Done. Ready to get merged?

StephanEwen commented on flink-java/src/main/java/org/apache/flink/api/java/tuple/builder/Tuple0Builder.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Tuple0Builder {
+
+	private List<Tuple0> tuples = new LinkedList<Tuple0>();
```
ArrayLists are almost always way superior in performance than LinkedLists.

StephanEwen replied:
And memory efficiency...

mjsax replied:
I copied from Tuple1Builder, ...Tuple25Builder. Will introduce ArrayList for all other, too.

StephanEwen commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/runtime/Tuple0Serializer.java:
```diff
@@ -0,0 +1,86 @@
+
+	@Override
+	public Tuple0Serializer duplicate() {
+		return new Tuple0Serializer();
```
No need to create a new instance, since this is stateless.

StephanEwen commented on flink-staging/flink-language-binding/flink-language-binding-generic/src/main/java/org/apache/flink/languagebinding/api/java/common/streaming/Receiver.java:
```diff
@@ -346,61 +349,12 @@ public Tuple deserialize() {
-			default:
-				throw new IllegalArgumentException("Tuple size not supported: " + size);
+		try {
+			return Tuple.getTupleClass(size).newInstance();
```
This should have a size check and give a proper error message.

mjsax replied:
No need for this. `.getTupleClass()` does the check already.

StephanEwen replied:
Yep, you're right...

StephanEwen commented:
Some comments inline, other than that, two issues with this pull request:
  - A lot of whitespace reformatting. We explicitly ask not to do this. Some IDEs do it automatically, but you can deactivate it. It makes diffs dangerously convoluted.
  - Tuple0 can be treated as a singleton, since it has no state. Any reason not to do this?

mjsax commented:
- changed `LinkedList` to `ArrayList` in **all** `TupleXBuilder` classes
- removed WS formatting
- changed `Tuple0` **and** `Tuple0Serializer` to singleton

StephanEwen commented:
Two comments on the singleton change:
  - The Tuple0.class.newInstance() will now probably fail (due to the private constructor)
  - Java serialization still creates copies. Need to overwrite the `readResolve()` method to create proper singletons.
  - It is probably okay allow Tuple0 instances to be created. We could simply providing a static shareable instance in the class for convenience (soft-singleton style).

@twalthr Are the TypeExctractor related changes good?

Thanks for reverting the whitespace changes. I did actually not expect that, it was more a comment for future pull requests...

mjsax commented:
You are right about `newInstance`. I removed private constructor. Soft singleton is used now. Also added `readResolve()` (something new I learned today :) ).

StephanEwen commented:
Was it on purpose to add `readResolve()` to the serializer, rather than the tuple?

mjsax commented:
Ups. That was not intended at all... Does not make any sense. Sorry. I didn't realize that I edited the wrong class because Tuple0Serializer is singleton, too... Just fixed it.

mjsax commented:
Any news about this PR?  @twalthr : Are you going to review it again?

twalthr commented:
The typeutil classes look good. I see that you have modified the TupleXXBuilders, have you modified them by hand or by running the `TupleGenerator`? I can't see the modified `TupleGenerator` in your PR.

mjsax commented:
I modified by hand. Was not aware of `TupleGenerator`. Just had a look into it. Not sure if `Tuple0` can be included appropriately. For example, it is no generic class; it is implemented as soft singleton. Extending `TupleGenerator` would result in special handling of Tuple0 in every place. Thus, adding it manually and not generate the source code for it seems better to me.

StephanEwen commented:
Agree, Tuple0 should probably not go through the tuple generator.

twalthr commented:
Your are right Tuple0 support for the TupleGenerator is not important. Actually, I meant the change of `import java.util.LinkedList;` to `import java.util.ArrayList;` otherwise these changes get lost if the TupleGenerator is executed.

mjsax commented:
I did those manually, too. I am just reworking the code to fix this up.

mjsax commented:
I did the following changes:
 - added Tuple0 to `TupleGenerator.modifyTupleType()`
 - changes `LinkedList` to `ArrayList` in TupleGenerator to create code for `TupleX.java`
 - added methods `toString()`, `equals()`, and `hashCode()` to `Tuple0`
(Of course, I run TupleGenerator to test it.)

This PR should be ready for merging now.

StephanEwen commented:
Looks good, will merge this.

Will make slight adjustments in the merge (for code style consistency), like naming the single instance uppercase `INSTANCE`.
